src/hotspot/share/compiler/compileBroker.cpp

```diff
@@ -1106,6 +1111,8 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
 #endif
       JavaThread *ct = make_thread(compiler_t, compiler2_object(i), _c2_compile_queue, _compilers[1], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```

Comment:
This seems to be the fully dynamic case. Here the TLH is potentially racing with thread termination and so may or may not capture ct. Rather than an assertion the call to  `get_thread_name()` below should be guarded by `tlh.includes(ct)`.